prov/rxm/src/rxm_ep.c

```diff
@@ -422,82 +433,114 @@ static int rxm_ep_peek_recv(struct rxm_ep *rxm_ep, fi_addr_t addr, uint64_t tag,
 }
 
-static ssize_t rxm_ep_recv_common(struct rxm_ep *rxm_ep, const struct iovec *iov,
-				  void **desc, size_t count, fi_addr_t src_addr,
-				  uint64_t tag, uint64_t ignore, void *context,
-				  uint64_t flags, struct rxm_recv_queue *recv_queue)
+static inline ssize_t
+rxm_ep_format_receive(struct rxm_ep *rxm_ep, const struct iovec *iov,
+		      void **desc, size_t count, fi_addr_t src_addr,
+		      uint64_t tag, uint64_t ignore, void *context,
+		      uint64_t flags, struct rxm_recv_queue *recv_queue,
+		      struct rxm_recv_entry **recv_entry)
+{
+	size_t i;
+
+	*recv_entry = rxm_recv_entry_get(recv_queue);
+	if (OFI_UNLIKELY(!*recv_entry))
+		return -FI_EAGAIN;
+
+	(*recv_entry)->count 	= (uint8_t)count;
+	(*recv_entry)->addr 	= RXM_EP_RECV_SANITIZE_SRC_ADDR(rxm_ep, src_addr);
+	(*recv_entry)->context 	= context;
+	(*recv_entry)->flags 	= flags;
+	(*recv_entry)->ignore 	= ignore;
+	(*recv_entry)->tag	= tag;
+
+	for (i = 0; i < count; i++) {
+		(*recv_entry)->iov[i].iov_base = iov[i].iov_base;
+		(*recv_entry)->iov[i].iov_len = iov[i].iov_len;
+		if (desc)
+			(*recv_entry)->desc[i] = desc[i];
+	}
+
+	return FI_SUCCESS;
+}
+
+static inline ssize_t
+rxm_ep_recv_common(struct rxm_ep *rxm_ep, const struct iovec *iov,
+		   void **desc, size_t count, fi_addr_t src_addr,
+		   uint64_t tag, uint64_t ignore, void *context,
+		   uint64_t op_flags, struct rxm_recv_queue *recv_queue)
 {
 	struct rxm_recv_entry *recv_entry;
 	struct rxm_rx_buf *rx_buf;
-	size_t i;
+	ssize_t ret;
 
 	assert(count <= rxm_ep->rxm_info->rx_attr->iov_limit);
 
-	if (flags & (FI_PEEK | FI_CLAIM | FI_DISCARD))
-		assert(recv_queue->type == RXM_RECV_QUEUE_TAGGED);
+	ret = rxm_ep_format_receive(rxm_ep, iov, desc, count, src_addr,
+				    tag, ignore, context, op_flags,
+				    recv_queue, &recv_entry);
+	if (OFI_UNLIKELY(ret))
+		return ret;
+
+	fastlock_acquire(&recv_queue->lock);
+	rx_buf = rxm_check_unexp_msg_list(recv_queue, src_addr, tag,
+					  ignore);
+	if (rx_buf) {
+		dlist_remove(&rx_buf->unexp_msg.entry);
+		rx_buf->recv_entry = recv_entry;
+		fastlock_release(&recv_queue->lock);
+		return rxm_cq_handle_data(rx_buf);
+	}
+
+	RXM_DBG_ADDR_TAG(FI_LOG_EP_DATA, "Enqueuing recv", recv_entry->addr,
+			 recv_entry->tag);
+	dlist_insert_tail(&recv_entry->entry, &recv_queue->recv_list);
+	fastlock_release(&recv_queue->lock);
+	return 0;
+}
+
+static ssize_t rxm_ep_recv_common_flags(struct rxm_ep *rxm_ep, const struct iovec *iov,
+					void **desc, size_t count, fi_addr_t src_addr,
+					uint64_t tag, uint64_t ignore, void *context,
+					uint64_t flags, uint64_t op_flags,
+					struct rxm_recv_queue *recv_queue)
+{
+	struct rxm_recv_entry *recv_entry;
+	struct rxm_rx_buf *rx_buf;
 
-	src_addr = (rxm_ep->rxm_info->caps & FI_DIRECTED_RECV) ?
-		src_addr : FI_ADDR_UNSPEC;
+	assert(count <= rxm_ep->rxm_info->rx_attr->iov_limit);
+	assert((recv_queue->type == RXM_RECV_QUEUE_TAGGED) &&
+		((recv_queue->type != RXM_RECV_QUEUE_TAGGED) &&
```

Comment:
Doesn't this throw an assertion failure always?

Reply:
How about `assert(!(flags & (FI_PEEK | FI_CLAIM | FI_DISCARD)) || (recv_queue->type == RXM_RECV_QUEUE_TAGGED));`?

Reply:
I've checked your proposals w/ debug build. It works fine. Thanks! 👍